<img src="assets/image_2.jpeg" width="350" style="margin-left:auto; margin-right:auto" />

<h1><center>Project: Hacker News Posts</center></h1>

This analysis entails a comparative analysis of two distinct categories of posts on Hacker News. Therefore we will explore the Hacker News [dataset](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts). Regarding the information on kaggle, this dataset contains Hacker News posts for 12 month, up to September 26 2016. The Data is coming from the Hacker News [website](https://news.ycombinator.com). 

##### What is Hacker News?
Hacker News is a prominent platform where technology-related stories, referred to as 'posts,' undergo voting and commentary. The two categories under examination are delineated by posts commencing with either `'Show HN'` or `'Ask HN'`. `'Show HN'` posts are submissions aimed at showcasing projects, products, or intriguing content to the Hacker News community. Conversely, `'Ask HN'` posts are submissions directed at the Hacker News community to pose specific inquiries, such as soliciting recommendations for exemplary online courses.

##### What is the project about?
Our specific focus involves a comparative investigation between these two post types, aiming to discern:
- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

It is to note that the dataset utilized underwent a reduction process, wherein nearly 300,000 rows were initially narrowed down to approximately 20,000 rows. This reduction involved the elimination of submissions lacking any comments, followed by a random sampling of the remaining submissions.
##### The dataset
The dataset includes the following columns:
- **ID**: The ID of the user
- **title**: title of the post
- **url**: the url of the item being linked to
- **num_points**: the number of upvotes the post received
- **num_comments**: the number of comments the post received
- **author**: the name of the account that made the post
- **created_at**: the date and time the post was made (the time zone is Eastern Time in the US)

Lets start.

*This project was completed as part of the Data Science Career Path offered by dataquest.io.*

#### The Data

In [1]:
from csv import reader
import datetime as dt

# Read in the data.
opened_file = open('hacker_news.csv')
hn = list(reader(opened_file))

In [2]:
hn[:5] # print the first 5 rows of hn (hacker_news dataset)

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

#### Removing Headers from List of Lists

In [3]:
headers = hn[0] #storing the header row into a separate variable
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [4]:
hn = hn[1:] #storing the data without header into the hn variable
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

The dataset encompasses post titles, the corresponding count of comments per post, and the posting date. Our initial exploration will delve into examining the comment counts categorized by post types.

#### Extracting Posts by Post Types

In the steps below, we will discern posts commencing with either `'Ask HN'` or `'Show HN'` and subsequently partition the dataset into distinct lists based on these two post types.

In [5]:
# Identify posts initiated by either 'Ask HN' or 'Show HN', subsequently segmenting the dataset into distinct lists based on these defined post types
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    
    if title.lower().startswith('ask hn'): #using startswith comment to separate ask and show posts
        ask_posts.append(row)
    elif title.lower().startswith('show hn'): 
        show_posts.append(row)
    else:
        other_posts.append(row)

In [6]:
print(len(ask_posts)) #3488 posts with ask hn
print(len(show_posts)) #1162 posts with show hn
print(len(other_posts)) # 17193 other posts, without ask or show hn

1744
1162
17194


#### Determining the Mean Comment Count for 'Ask HN' and 'Show HN' Posts

We will proceed to compute the mean comment count garnered by each category of posts.

In [7]:
# Calculate the average number of comments `Ask HN` posts receive
total_ask_comments = 0

for row in ask_posts:
    comments = int(row[4])
    if comments != 0:
        total_ask_comments += comments

avg_ask_comments = total_ask_comments / len(ask_posts)

print(avg_ask_comments) #avg comments on ask posts ~14

14.038417431192661


In [8]:
total_show_comments = 0


for row in show_posts:
    comments = int(row[4])
    if comments != 0:
        total_show_comments += comments
        
avg_show_comments = total_show_comments / len(show_posts)

print(avg_show_comments) #avg comment on show posts ~10.

10.31669535283993


Following the categorization into the `'Show Post'` and `'Ask Post'` sections and a comprehensive analysis of each segment, it is evident that a post in the `'Ask Post'` category typically garners approximately 14 comments on average, whereas a post in the `'Show Post'` category tends to accumulate around 10 comments on average.

#### Determining the Amount of Ask Posts and Corresponding Comments by Hour of Creation

Next, we aim to determine if there is a optimal timing to maximize the amount of comments garnered by `'Ask posts'`. Initially, we will determine the quantity of `'Ask posts'` generated during each hour of the day, along with the count of comments these posts received. Subsequently, we will compute the average count of comments received by `'Ask posts'` created at each hour throughout the day.

In [9]:
# First: Create a list with date and number of comments per date
result_list = []

for row in ask_posts:
    created_at = row[6]
    comments = int(row[4])
    result_list.append([created_at, comments])

# Second: Calculate the amount of ask posts created during each hour of day and the number of comments received
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    created_date = row[0]
    comment = row[1]
    date = dt.datetime.strptime(created_date, '%m/%d/%Y %H:%M')
    hour = date.strftime('%H')
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment

print('Amount of ask posts by hour: ' + str(counts_by_hour)) 
print('\n')
print('Amount of comments by hour: '  + str(comments_by_hour))

Amount of ask posts by hour: {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Amount of comments by hour: {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [10]:
max(counts_by_hour.values()) # Maximum value for amount of post by hour is 116 at 3 p.m

116

In [11]:
max(comments_by_hour.values()) # Maximum value for comments by hour is 4477 at 3 p.m

4477

#### Calculating the Average Number of Comments for Ask HN Posts by Hour

In [12]:
# Compute the mean quantity of comments that 'Ask HN' posts, generated at each hour of the day, receive
avg_by_hour = []

for comment in comments_by_hour:
    avg_by_hour.append([comment, round(comments_by_hour[comment] / counts_by_hour[comment],1)])

avg_by_hour

[['09', 5.6],
 ['13', 14.7],
 ['10', 13.4],
 ['14', 13.2],
 ['16', 16.8],
 ['23', 8.0],
 ['12', 9.4],
 ['17', 11.5],
 ['15', 38.6],
 ['21', 16.0],
 ['20', 21.5],
 ['02', 23.8],
 ['18', 13.2],
 ['03', 7.8],
 ['05', 10.1],
 ['19', 10.8],
 ['01', 11.4],
 ['22', 6.7],
 ['08', 10.2],
 ['04', 7.2],
 ['00', 8.1],
 ['06', 9.0],
 ['07', 7.9],
 ['11', 11.1]]

#### Sorting and Printing Values from a List of Lists

In [13]:
switch_avg_by_hour = []

for element in avg_by_hour:
    switch_avg_by_hour.append([element[1], element[0]])
    
switch_avg_by_hour

[[5.6, '09'],
 [14.7, '13'],
 [13.4, '10'],
 [13.2, '14'],
 [16.8, '16'],
 [8.0, '23'],
 [9.4, '12'],
 [11.5, '17'],
 [38.6, '15'],
 [16.0, '21'],
 [21.5, '20'],
 [23.8, '02'],
 [13.2, '18'],
 [7.8, '03'],
 [10.1, '05'],
 [10.8, '19'],
 [11.4, '01'],
 [6.7, '22'],
 [10.2, '08'],
 [7.2, '04'],
 [8.1, '00'],
 [9.0, '06'],
 [7.9, '07'],
 [11.1, '11']]

Below we'll switch the values of the list, therefore the count will be the first element of the list and the hour the second.

In [14]:
sorted_switch = sorted(switch_avg_by_hour, reverse=True)
sorted_switch[:5]

[[38.6, '15'], [23.8, '02'], [21.5, '20'], [16.8, '16'], [16.0, '21']]

#### The 5 Hours with the Highest Average Comments

In [15]:
# Sort the values and print the the 5 hours with the highest average comments.
from datetime import timedelta

for element in sorted_switch[:5]:
    template = "{time} o'clock -> {avg_posts} average comments per post."
    avg_posts = element[0]
    time = dt.datetime.strptime(element[1], '%H')
    time = time - timedelta(hours=5) #correcting the time zone, for Central European Standard Time
    time_str = time.strftime('%H:%M')
    
    output = template.format(time=time_str, avg_posts=avg_posts)
    print(output)

10:00 o'clock -> 38.6 average comments per post.
21:00 o'clock -> 23.8 average comments per post.
15:00 o'clock -> 21.5 average comments per post.
11:00 o'clock -> 16.8 average comments per post.
16:00 o'clock -> 16.0 average comments per post.


Looking at the results, you can see, that posting into the `'Ask posts'` section at 10 o'clock will give you the most comments for your post on average *38.6*. The second most comments on average (*23.8*) are at 21 o'clock. There exists an approximate 60% surge in the comment count between the hours exhibiting the highest and second-highest average numbers of comments.

#### Determine if `'Show posts'` or `'Ask posts'` Receive More Points On Average

In [16]:
# Calculate the average number of points `Ask HN` posts receive
ask_posts_point = 0

for row in ask_posts:
    points = int(row[3])
    
    if points != 0:
        ask_posts_point += points
        
avg_ask_points = ask_posts_point / len(ask_posts)   

avg_ask_points #15.1 on average

15.061926605504587

In [17]:
# Calculate the average number of points `Show HN` posts receive
show_posts_point = 0

for row in show_posts:
    points = int(row[3])
    
    if points != 0:
        show_posts_point += points
        
avg_show_points = show_posts_point / len(show_posts)  

avg_show_points # 27.5 on average

27.555077452667813

As depicted above, posts crafted within the `'Show posts'` category exhibit a notably higher average point count (27.5) in contrast to the `'Ask posts'` category (15.1). This distinction might be attributed to the nature of `'Show posts'` that typically showcase user-generated content, often representing noteworthy accomplishments. Consequently, users tend to manifest their commendation through points, thereby contributing to the higher average points observed within the `'Show posts'` section compared to the `'Ask posts'` section. 

In further analysis, we will concentrate on the `'Show posts'` section characterized by elevated point counts, exploring whether specific posting times correlate with higher point reception.

#### Show Posts Data

In [18]:
show_posts[:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

In [19]:
# Save the creation date and the points into a list
result_list_show = []

for row in show_posts:
    created_at = row[6] #date and time of posting
    points = int(row[3]) #points
    result_list_show.append([created_at, points])

In [20]:
# Save the creation hour and points/count into a dictionary
counts_by_hour_show = {}
points_by_hour = {}

for row in result_list_show:
    created_at = row[0]
    points = row[1]
    date = dt.datetime.strptime(created_at, '%m/%d/%Y %H:%M')
    hour = date.strftime('%H')
    
    if hour not in counts_by_hour_show:
        counts_by_hour_show[hour] = 1
        points_by_hour[hour] = points
    else:
        counts_by_hour_show[hour] += 1
        points_by_hour[hour] += points

In [21]:
# Save the average points by hour into a list
avg_by_hour_show = []

for hour in points_by_hour:
    avg_by_hour_show.append([hour, round((points_by_hour[hour] / counts_by_hour_show[hour]),1)])

In [22]:
# switch the values and sort them by average points
switch_avg_by_hour_show = []

for point in avg_by_hour_show:
    switch_avg_by_hour_show.append([point[1], point[0]])

sorted_switch_show = sorted(switch_avg_by_hour_show, reverse=True)
sorted_switch_show

[[42.4, '23'],
 [41.7, '12'],
 [40.3, '22'],
 [37.8, '00'],
 [36.3, '18'],
 [33.6, '11'],
 [30.9, '19'],
 [30.3, '20'],
 [28.6, '15'],
 [28.3, '16'],
 [27.1, '17'],
 [25.4, '14'],
 [25.1, '03'],
 [25.0, '01'],
 [24.6, '13'],
 [23.4, '06'],
 [19.0, '07'],
 [18.9, '10'],
 [18.4, '21'],
 [18.4, '09'],
 [15.3, '08'],
 [14.8, '04'],
 [11.3, '02'],
 [5.5, '05']]

In [23]:
for element in sorted_switch_show[:5]:
    avg = element[0]
    
    hours = dt.datetime.strptime(element[1], '%H')
    hours = hours - timedelta(hours=5) #adjust for Central European Standard Time
    hours_str = hours.strftime('%H:%M')
    
    template = "At {hours_str} o'clock were {avg} points on average at the show post section"
    output = template.format(avg=avg, hours_str=hours_str)

    print(output)

At 18:00 o'clock were 42.4 points on average at the show post section
At 07:00 o'clock were 41.7 points on average at the show post section
At 17:00 o'clock were 40.3 points on average at the show post section
At 19:00 o'clock were 37.8 points on average at the show post section
At 13:00 o'clock were 36.3 points on average at the show post section


We executed a similar procedure for the `'Show Posts'` section as previously conducted for the `'Ask Posts'` section. However, in this instance, our analysis focused on determining the average points per post within the `'Show Posts'` section, diverging from the earlier examination that centered on the average comments per post within the `'Ask Posts'` section. 

Presented above are the top five instances where the average points per post in the `'Show Posts'` section reached peak values. Specifically, at 18:00 o'clock, the average points per post approximate 42.4. An early morning post (Central European Standard Time) at 7:00 o'clock garnered the second-highest average points per post (41.7).


#### The `'Other Posts`' Section

Having scrutinized the average comments within the `'Ask Posts'` section and the average points within the `'Show Posts'` section, our subsequent step involves an analysis of the `'Other Posts'` section. This examination aims to establish a comparative assessment against the figures observed in the `'Ask Posts'` and `'Show Posts'`  sections.

#### Other Posts Data

In [24]:
other_posts[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [25]:
other_posts_point = 0

for row in other_posts[1:]:
    points = int(row[3])
    
    if points != 0:
        other_posts_point += points
        
avg_other_post_points = other_posts_point / len(other_posts)  

In [26]:
other_posts_comment = 0

for row in other_posts[1:]:
    comment = int(row[4])
    
    if comment != 0:
        other_posts_comment += comment
        
avg_other_posts_comment = other_posts_comment / len(other_posts)     

In [27]:
print(avg_other_post_points)
print(avg_other_posts_comment)

55.38432011166686
26.870012795161102


The average point count for posts categorized under `'Other Posts'` stands notably higher at ~55.4 compared to the average point counts in the `'Show Posts'` (27.5) and `'Ask Posts'` (15.1) sections. Additionally, the average comment count for `'Other Posts'` (26.9) significantly surpasses the approximately 14 average comments per post in the 'ask' section and the approximately 10 average comments per post in the 'show' section. 

For further investigation, we will delve deeper into the instances when the average comments and points reach their peaks.

In [28]:
result_list_other_points = []
result_list_other_comments = []

for row in other_posts[1:]:
    created_at = row[6] #date and time of posting
    points = int(row[3]) #points count
    comments = int(row[4]) #comments count
    result_list_other_points.append([created_at, points])
    result_list_other_comments.append([created_at, comments])

First, we'll create the two dictionaires for the points count.

In [29]:
counts_by_hour_other= {}
points_by_hour_other = {}

for row in result_list_other_points:
    created_at = row[0]
    points = row[1]
    date = dt.datetime.strptime(created_at, '%m/%d/%Y %H:%M')
    hour = date.strftime('%H')
    
    if hour not in counts_by_hour_other:
        counts_by_hour_other[hour] = 1
        points_by_hour_other[hour] = points
    else:
        counts_by_hour_other[hour] += 1
        points_by_hour_other[hour] += points

Next, we'll create the two dictionaires for the comments count.

In [30]:
counts_by_hour_other_comments= {}
by_hour_other_comments = {}

for row in result_list_other_comments:
    created_at = row[0]
    comments = row[1]
    date = dt.datetime.strptime(created_at, '%m/%d/%Y %H:%M')
    hour = date.strftime('%H')
    
    if hour not in counts_by_hour_other_comments:
        counts_by_hour_other_comments[hour] = 1
        by_hour_other_comments[hour] = comments
    else:
        counts_by_hour_other_comments[hour] += 1
        by_hour_other_comments[hour] += comments

#### Data Processing Like Ask Posts and Show Posts

In [31]:
points_avg_by_hour_other = []

for hour in points_by_hour_other:
    points_avg_by_hour_other.append([hour, round((points_by_hour_other[hour] / counts_by_hour_other[hour]),1)])

comments_avg_by_hour_other = []

for hour in by_hour_other_comments:
    comments_avg_by_hour_other.append([hour, round((by_hour_other_comments[hour] / counts_by_hour_other_comments[hour]),1)])

In [32]:
switch_avg_by_hour_other_points = []

for point in points_avg_by_hour_other:
    switch_avg_by_hour_other_points.append([point[1], point[0]])
switch_avg_by_hour_other_points = sorted(switch_avg_by_hour_other_points, reverse=True)

switch_avg_by_hour_other_comments = []

for comment in comments_avg_by_hour_other:
    switch_avg_by_hour_other_comments.append([comment[1], comment[0]])
switch_avg_by_hour_other_comments = sorted(switch_avg_by_hour_other_comments, reverse=True)

#### Other Posts Results: Points and Comments

In [33]:
for row in switch_avg_by_hour_other_points[:5]:
    avg_points = row[0]
    time = dt.datetime.strptime(row[1], '%H')
    time = time - timedelta(hours=5)
    hour_str = time.strftime('%H:%M')
    
    template = "At {hour_str} o'clock, there are {avg_points} points on average in the other post section"
    output = template.format(hour_str=hour_str, avg_points=avg_points)
    print(output)

At 08:00 o'clock, there are 62.5 points on average in the other post section
At 09:00 o'clock, there are 61.8 points on average in the other post section
At 10:00 o'clock, there are 60.5 points on average in the other post section
At 05:00 o'clock, there are 60.5 points on average in the other post section
At 14:00 o'clock, there are 60.0 points on average in the other post section


In [34]:
for row in switch_avg_by_hour_other_comments[:5]:
    avg_comments = row[0]
    time = dt.datetime.strptime(row[1], '%H')
    time = time - timedelta(hours=5)
    hour_str = time.strftime('%H:%M')
    
    template = "At {hour_str} o'clock, there are {avg_comments} comments on average in the other post section"
    output = template.format(hour_str=hour_str, avg_comments=avg_comments)
    print(output)

At 09:00 o'clock, there are 32.3 comments on average in the other post section
At 08:00 o'clock, there are 30.9 comments on average in the other post section
At 07:00 o'clock, there are 30.3 comments on average in the other post section
At 06:00 o'clock, there are 29.6 comments on average in the other post section
At 10:00 o'clock, there are 29.5 comments on average in the other post section


The segment denoted as `'Other Posts'` exhibits the peak average points at 8 a.m., reaching 62.5. Simultaneously, the apex average comment count within the `'Other Posts'` section is observed at 9 a.m., totaling approximately 32 comments. The `'Other Posts'` section exhibits the highest average values for comments and points. This trend could be attributed to its broader scope, fostering more extensive discussions on diverse topics. Conversely, the `'Ask Posts'` primarily revolve around question-related topics, while the `'Show Posts'` predominantly feature user-generated projects and related discussions.

#### Conclusion


In this project, we conducted an analysis encompassing the `'Ask Posts'`, `'Show Posts'`, and `'Other Posts'` sections extracted from the Hacker News website. The objective was to identify the post types and corresponding time frames that tend to accumulate the highest average comments and points. Based on our comprehensive analysis, with a specific focus on the `'Ask Posts'` and `'Show Posts'` sections, our recommendation is to create a post in the `'Ask Post'` section, between 10:00 and 11:00 (Central European Standard Time), to achieve the highest average comment count. For optimizing average points, our suggestion is to create a post in the `'Show Post'` and creating it between 18:00 and 19:00 (Central European Standard Time).

However, it is crucial to acknowledge that our analysis was conducted on a dataset that excluded posts lacking any comments. Consequently, a more accurate statement would be that among the posts that received comments, the average comment count was higher for `'Ask Posts'` and the average point count was higher for `'Show Posts'` (excluding the `'Other Posts'` section).